basics/uri.adoc

```diff
@@ -25,404 +26,522 @@ specific language governing permissions and limitations
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                                                               |
-|              UAUTHORITY_ADDRESS (IPv6 128bits)                |
-|                                                               |
-|                                                               |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-----
+Similarly, a Rust library might implement a `UriValidator` struct providing corresponding functions
 
-.ID Remote Micro URI
-[#vin-micro-uri,java]     
+[source,rust]
 ----
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|  UP_VERSION   |      TYPE     |           URESOURCE_ID        |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|        UENTITY_ID             |  UE_VERSION   |   UNUSED      |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|    ID_LEN     |        UAUTHORITY_ID (1=256 bytes)  ...       |
-|                                                               |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+pub struct UriValidator {}
+
+impl UriValidator {
+  pub fn is_rpc_method(uri: &UUri) -> bool {
+    ...
+  }
+  pub fn validate_rpc_method(uri: &UUri) -> Result<(), UuriValidationError> {
+    if is_rpc_method(uri) {
+      Ok(())
+    } else {
+      Err(UuriValidationError::new("not an RPC Method URI"))
+    }
+  }
+}
 ----
 
-== URI Examples
-
-The following examples are based off the USubscription notification topic (see usubscription.proto).
-
-
-.Local Uri
-[width="80%",cols="30%,70%",options="header"]
-|===
-|Format | Value
-
-|Long
-|`/core.usubscription/3/SubscriptionChange#Update`
-
-|Short
-| `/0/3/32768`
-
-|Micro
-|`[1, 0, -128, 0, 0, 0, 3, 0]`
-
-|===
+or implement the functions on the `UUri` struct
 
+[source,rust]
+----
+impl UUri {
+  pub fn is_rpc_method(&self) -> bool {
+    ...
+  }
+  pub fn validate_rpc_method(&self) -> Result<(), UuriValidationError> {
+    if self.is_rpc_method() {
+      Ok(())
+    } else {
+      Err(UuriValidationError::new("not an RPC Method URI"))
+    }
+  }
+}
+----
 
-=== Remote Uris
+== Pattern Matching
 
+A UUri can be used to define a _pattern_ that other UUris can then be matched against.
 
-.Remote IPv4 Uri
-[width="80%",cols="30%,70%",options="header"]
-|===
-|Format | Value
+[.specitem,oft-sid="dsn~pattern-matching~1",oft-needs="impl,utest"]
+--
+A UUri that is used as a pattern to match other UUris against
 
-|Long
-|`//vcu.veh.gm.com/core.usubscription/3/SubscriptionChange#Update`
+* *MAY* have its `authority_name` set to the `*` (`U+002A`, Asterisk) character in order to match _any_ (including _no_) authority,
+* *MAY* have the _service type_ part of its `ue_id` set to `0xFFFF` in order to match _any_ type,
```

Comment:
maybe we need to call it service number to differentiate the instance id?

Reply:
not sure what you mean, replace _service type_ with _service number_ or replace _service instance_ with _service number_?

Reply:
Re-reading some/ip spec for a little inspiration, they have service_id and service_instance_id, no where do we refer to service_id as type hence why I wanted to keep the id or number but not type.

Reply:
then let's stick to _service id_ and service instance id_. I'll update accordingly ...